internal/controllers/gateway/udproute_controller.go

```diff
@@ -312,14 +312,28 @@ func (r *UDPRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		}
 	}
 
-	// TODO: UDPRoute should be 'Accepted' before proceeding further. Get here back once
-	// https://github.com/Kong/kubernetes-ingress-controller/issues/2544 is implemented.
-
-	// if the gateways are ready, and the UDPRoute is destined for them, ensure that
-	// the object is pushed to the dataplane.
-	if err := r.DataplaneClient.UpdateObject(udproute); err != nil {
-		debug(log, udproute, "failed to update object in data-plane, requeueing")
-		return ctrl.Result{}, err
+	if isRouteAccepted(gateways) {
+		// if the gateways are ready, and the UDPRoute is destined for them, ensure that
+		// the object is pushed to the dataplane.
+		if err := r.DataplaneClient.UpdateObject(udproute); err != nil {
+			debug(log, udproute, "failed to update object in data-plane, requeueing")
+			return ctrl.Result{}, err
+		}
+		if r.DataplaneClient.AreKubernetesObjectReportsEnabled() {
+			// if the dataplane client has reporting enabled (this is the default and is
+			// tied in with status updates being enabled in the controller manager) then
+			// we will wait until the object is reported as successfully configured before
+			// moving on to status updates.
+			if !r.DataplaneClient.KubernetesObjectIsConfigured(udproute) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+		}
+	} else {
+		// route is not accepted, remove it from kong store
+		if err := r.DataplaneClient.DeleteObject(udproute); err != nil {
+			debug(log, udproute, "failed to delete object in data-plane, requeueing")
+			return ctrl.Result{}, err
+		}
 	}
```

Comment:
Given we've got the envtests setup ready to be used, maybe we could try to cover the logic we're adding to reconcilers in such tests? Doesn't have to be done in this PR though.

Reply:
Do we have an example elsewhere of ripping things out of integration and moving them into smaller envtests? The "basics" integration tests have gotten rather long, with a running theme of:

- Resource starts in state (e.g. missing the now infamous unmanaged annotation.
- Resource event runs through controller.
- Controller changes state and requeues resource.

Integration as written takes a "run through a 'natural' set of state changes in order" approach, with ever more state change conditions getting tacked onto the end. Does envtest let us break that into components where we can assume that the Gateway resource is already programmed or whatever?

Integration needs to instantiate the Gateway and wait for the separate Gateway controller to do its thing before we can check UDPRoute state transitions. Does envtest let us provide a ready Gateway with valid Listeners independent of actually having the Kong instance and controller activity that would bring a Gateway to that state?

Reply:
I don't think we have an example of moving test logic from integration to envtests as the ones that @pmalek introduced were brand new: 
- https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/controllers/gateway/httproute_controller_envtest_test.go
- https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/controllers/configuration/kongadminapi_controller_envtest_test.go

It's possible to set a Gateway's state in those explicitly in them and have parametrized test cases based on that I think. There's no need to go through the whole "natural"/e2e flow to trigger a given state since envtests are running against a single reconciler only.

Reply:
What @czeslavo wrote + you can just spin up what you need (`UDPRouteReconciler` in this case) and prepare the test environment just like you would for any other test but adjust the preconditions, i.e. the UDPRouteReconciler might require a Gateway to have some predefined conditions etc. so just create them using the envtest environment just like in the tests linked by @czeslavo and then perform the test logic and the assertions you want.

Reply:
Got it. I think I'm sticking with integration for this one since this does need to test interactions betwen the Gateway and UDPRoute controllers, not just state changes in the latter, and because this needs to test Gateway controller reactions to the environment. We do want a more E2E flow here beyond just injecting a complete Gateway state for the UDPRoute controller. 

If we want UDPRoute envtests that do check behavior with an injected Gateway, I'd say schedule that in a follow-up. We can do it independently of the controller code.